[espnow] Add transport platform for packet_transport

EasilyBoredEngineer opened the pull request:
## Description
Adds ESP-NOW as a transport platform for the packet_transport component, enabling wireless communication between ESP32 devices. 

## Configuration

The transport supports both broadcast and unicast modes:

- **Broadcast mode** (default): Sends to all peers using `FF:FF:FF:FF:FF:FF`
- **Unicast mode**: Sends to a specific peer by setting `peer_address`

Peers must still be registered with the ESP-NOW component. The transport 
receives from all peers regardless of the peer_address setting.

## Example configuration

Please see Yaml below.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5458

## Checklist:
- [x] The code change is tested and works locally
- [x] Tests have been added to verify that the new code works (under `tests/` folder)
- [x] Code follows ESPHome style guidelines
- [x] Component added to CODEOWNERS

## Example YAML:
```yaml
esphome:
  name: test-espnow

esp32:
  board: esp32dev
  framework:
    type: esp-idf

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

logger:

espnow:
  id: espnow_component

packet_transport:
  - platform: espnow
    id: transport_broadcast
    espnow_id: espnow_component
    # Uses default broadcast address FF:FF:FF:FF:FF:FF
    encryption:
      key: "0123456789abcdef0123456789abcdef"
    sensors:
      - temp_sensor

  - platform: espnow
    id: transport_unicast
    espnow_id: espnow_component
    peer_address: "AA:BB:CC:DD:EE:FF"
    encryption:
      key: "0123456789abcdef0123456789abcdef"
    sensors:
      - temp_sensor

sensor:
  - platform: internal_temperature
    id: temp_sensor
    name: "Test Temperature"

EasilyBoredEngineer committed:
Add files via upload

EasilyBoredEngineer committed:
Create espnow-transport-test.yaml

test to use espnow with packet_transport

EasilyBoredEngineer committed:
Update CODEOWNERS

Add ESP-NOW transport platform for packet_transport component

Implements ESP-NOW as a transport option for the packet_transport component,
enabling wireless sensor data exchange between ESP32 devices using ESP-NOW
protocol with optional encryption support.

Features:
- Broadcast and unicast support
- Integration with existing packet_transport infrastructure
- Configurable broadcast address
- Automatic handler registration

EasilyBoredEngineer committed:
Update espnow-transport-test.yaml

improve test

EasilyBoredEngineer committed:
Add files via upload

change to unix style files

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Add files via upload

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

nielsnl68 reviewed (changes requested):
Hello @EasilyBoredEngineer,

Thanks for taking the time to add this component, could you please check the code around the current espnow components way to handle addresses for actions etc.

EasilyBoredEngineer committed:
Merge branch 'esphome:dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Add files via upload

address @nielsenlsn comments

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Add files via upload

Align ESP-NOW transport with ESP-NOW component address handling patterns

Update address handling in the ESP-NOW packet transport platform to match
the conventions used by the parent ESP-NOW component:

- Replace individual byte parameters with peer_address_t type
- Use std::array<uint8_t, 6> for broadcast address storage
- Update Python code to pass MAC address as HexInt array
- Maintain consistency with ESP-NOW component's MAC address API

This change ensures the transport platform follows established patterns
in the ESP-NOW ecosystem for handling MAC addresses, improving code
consistency and maintainability.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer commented:
Here's a ready-to-review message you can post as a comment on your PR:

---

## Ready for Review

This PR is now ready for maintainer review. All requested changes have been implemented:

### Changes Made

✅ **Address handling updated** - Replaced individual byte parameters with `peer_address_t` type to match the ESP-NOW component's established patterns for MAC address handling

✅ **Code formatting** - All linting issues resolved by pre-commit.ci bot

✅ **Test configuration** - Added test YAML configuration in `tests/components/espnow/test.espnow.yaml`

### Current Status

- Component compiles successfully
- All formatting checks pass
- Tested working with real hardware (ESP32-C3 devices successfully exchanging sensor data)
- Code follows ESPHome style guidelines
- Consistent with existing ESP-NOW component patterns

### Documentation

Documentation will be submitted as a separate PR to the `esphome-docs` repository once the code is approved, as discussed.

### Testing

The component has been tested with:
- ESP32-C3 devices
- Encryption enabled
- Sensor data transmission (internal temperature sensor)
- Both broadcast and peer-to-peer communication

Please let me know if any additional changes are needed.

---

This message clearly communicates that you've addressed the feedback, the PR is functional and tested, and you're ready for the next round of review. It also sets expectations about the documentation being handled separately.

EasilyBoredEngineer committed:
Add files via upload

added trailing space

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Add files via upload

remove trailing spaces, add empty line at end of each file

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Update CODEOWNERS

remove /

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Update ci.yml

EasilyBoredEngineer committed:
Add files via upload

remove unused sections init_py

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Update __init__.py

EasilyBoredEngineer committed:
Add files via upload

remove provider

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer commented:
@nielsnl68 from what I can see from looking at other components that use packet-transport, the hard work is done by the component itself. Nonetheless, I think you may be talking about the way I handle macs - that's been changed

EasilyBoredEngineer committed:
Update ci.yml

revert commit

EasilyBoredEngineer committed:
Add files via upload

Rename broadcast_address to peer_address and add send error logging

- More accurate naming since address can be broadcast OR unicast
- Add error logging to send callback for debugging
- Default remains FF:FF:FF:FF:FF:FF (broadcast)
- Update all references and documentation

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

nielsnl68 commented on esphome/components/espnow/packet_transport/__init__.py:
```diff
@@ -0,0 +1,39 @@
+CONFIG_SCHEMA = transport_schema(ESPNowTransport).extend(
+    {
+        cv.GenerateID(CONF_ESPNOW_ID): cv.use_id(ESPNowComponent),
+        cv.Optional(CONF_PEER_ADDRESS, default="FF:FF:FF:FF:FF:FF"): cv.mac_address,
```
One small thing. Using broadcast is a not preferably option. because you could probably interfere with other devices that are using espnow in your neighborhood.

EasilyBoredEngineer replied:
Could you clarify your concern about broadcast? My understanding is that ESP-NOW packets occupy the same airtime regardless of destination address (unicast vs broadcast) - the destination MAC only determines which device processes the packet, not the RF characteristics.

Are you concerned that:

- Broadcast encourages more devices to listen/respond, creating more bidirectional traffic?
- There's an RF-level difference I'm not aware of between unicast and broadcast ESP-NOW?
- Something else about broadcast behavior?

The Reality:

- ESP-NOW operates on WiFi channels (2.4GHz)
- ALL ESP-NOW packets are broadcast at the 802.11 MAC layer
- Every device in range "sees" the transmission regardless of destination address
- The destination MAC just determines which device processes it, not who receives it at the radio level

So the concern is:
Whether we send to FF:FF:FF:FF:FF:FF or AA:BB:CC:DD:EE:FF, it occupies the same airtime and creates the same interference. The destination address doesn't change the RF impact at all.

ESP-NOW's Built-in Protection:

- Peer Registration Required: Devices must usually (unless overridden in espnow setup with auto_add_peer=true) explicitly register peers.
- Optional Encryption: The ESP-NOW component supports encrypted communication. When encryption is enabled (via PMK/LMK), only devices with the correct keys can decrypt packets - even if they receive them.

Design Rationale:

The broadcast default was chosen because:

- It matches common ESP-NOW usage patterns (hub-and-spoke, sensor networks)
- It's the most flexible for initial setup and testing
- Users can easily override with peer_address for point-to-point scenarios
- It follows the principle: "secure by peer registration, flexible by default"

__
I want to make sure I address the right issue. Thanks!

nielsnl68 replied:
I'm fully aware how ESPNow works, i'm one of the creators of the espnow component, you know.

> Are you concerned that:

> - Broadcast encourages more devices to listen/respond, creating more bidirectional traffic?
> - There's an RF-level difference I'm not aware of between unicast and broadcast ESP-NOW?
> - Something else about broadcast behavior?

The issue I see with sending broadcast messages using espnow is that EVERY device in the near neighbourhood will receive the message. Not only yours but also other ESPNOW devices from other vendors. When that device has not an up-to-date or has bad designed firmware it could receive that message as if it is coming from a known sender and could crash when the data is interpreted the wrong way.

Sending messages using unicast (AA:BB:CC:DD:EE:FF) dont have that issue at all and is more secure.

> Peer Registration Required: Devices must usually (unless overridden in espnow setup with auto_add_peer=true) explicitly register peers.

Correct, but that works only on unicast (peer-to-peer) communication, and is half-way implemented. you will still receive messages from an unpaired device.

> Optional Encryption: The ESP-NOW component supports encrypted communication. When encryption is enabled (via PMK/LMK), only devices with the correct keys can decrypt packets - even if they receive them.

Please be aware that encryption is not supported yet by the ESPNow Component at the moment.

> It matches common ESP-NOW usage patterns (hub-and-spoke, sensor networks)

I am aware, and i find that a bad design solution. Until recently there was no way to determine if a received message was send directly or was send via broadcast. which could mess-up the incoming messages as descript above.

nielsnl68 replied:
Anyway, just let the end-user decide if (s)he want to broadcast or not.

swoboda1337 replied:
Its just a default can't the user change it to whatever they want?

EasilyBoredEngineer commented:
> Wow, thank you @EasilyBoredEngineer, that's a very neat addition which I was thinking to implement too, just a week ago 😅 I will be happy to test and contribute this weekend.

Thankyou :)

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Regenerate CODEOWNERS to fix bot warnings

EasilyBoredEngineer committed:
Update espnow-transport-test.yaml

changed broadcast_address to peer_address -- this option can be set unicast

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

swoboda1337 commented on esphome/components/espnow/packet_transport/espnow_transport.cpp:
```diff
@@ -0,0 +1,95 @@
+    return false;
+  }
+
+  std::vector<uint8_t> packet_data(data, data + size);
```
To avoid memory allocations it would be more efficient to make packet_data a member variable. Can do a resize and memcpy here.

swoboda1337 commented on esphome/components/espnow/packet_transport/espnow_transport.cpp:
```diff
@@ -0,0 +1,95 @@
+    return false;
+  }
+
+  std::vector<uint8_t> packet_data(data, data + size);
```
Can reuse the member variable here.

swoboda1337 commented on tests/components/espnow/espnow-transport-test.yaml:
```diff
@@ -0,0 +1,41 @@
+  ssid: !secret wifi_ssid
+  password: !secret wifi_password
+
+logger:
```
You shouldn't be adding the above yaml into the test file. If the file is named properly like test.esp32-idf.yaml that will added automatically.

swoboda1337 replied:
Can add the packet transport stuff get added to common.yaml?

swoboda1337 reviewed (changes requested):
Needs docs as well, see: 

https://esphome.io/components/packet_transport/
https://esphome.io/components/packet_transport/udp/
https://esphome.io/components/packet_transport/uart/

EasilyBoredEngineer committed:
Update espnow_transport.h

add packet_buffer_ vector - see https://github.com/esphome/esphome/pull/11025#discussion_r2414489038

EasilyBoredEngineer committed:
Update espnow_transport.cpp

reused packet_buffer_ -- 

see https://github.com/esphome/esphome/pull/11025/files/d8d912731cbdc268f62a242bf8d736de213a611b#r2414489038

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Add files via upload

Changes Made

✅ Added packet_buffer_ member variable to avoid memory allocations on every receive
✅ Updated both handlers to use resize() + memcpy() instead of vector constructor
✅ Fixed formatting - proper line length for the peer address log statement

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Delete tests/components/espnow/espnow-transport-test.yaml

delete epnow-transport yaml to move to subfolder

EasilyBoredEngineer committed:
Add files via upload

add packet_transport tests

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Add files via upload

remove trailing space

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

swoboda1337 commented on tests/components/espnow/packet_transport/common.yaml:
```diff
@@ -0,0 +1,34 @@
+
+wifi:
+  ssid: MySSID
+  password: password1
```
Take a look at: https://github.com/esphome/esphome/blob/dev/tests/test_build_components/build_components_base.esp32-idf.yaml

You shouldn't have `esphome:` here

EasilyBoredEngineer committed:
Update common.yaml

removed esphome and wifi per swabooda request.

swoboda1337 committed:
Merge remote-tracking branch 'upstream/dev' into espnow-packet-transport

swoboda1337 committed:
Fix tests

swoboda1337 committed:
Make some prints verbose and remove some

swoboda1337 commented:
Fixed the tests, you can't have sub folders the test wasn't actually run.

swoboda1337 commented on esphome/components/espnow/packet_transport/espnow_transport.cpp:
```diff
@@ -0,0 +1,95 @@
+
+  // Register this transport as an ESP-NOW handler
+  this->parent_->register_received_handler(static_cast<ESPNowReceivedPacketHandler *>(this));
+  this->parent_->register_broadcasted_handler(static_cast<ESPNowBroadcastedHandler *>(this));
```
Should these register calls be conditional? If broadcast is not used do we want to register for broadcast messages? Seems like a lot of extra overhead.

swoboda1337 replied:
@nielsnl68? anyone?

EasilyBoredEngineer committed:
Register broadcast handler for broadcast address

Conditionally register broadcast handler based on peer address.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

EasilyBoredEngineer committed:
Merge branch 'dev' into espnow-packet-transport

bdraco committed:
Merge branch 'dev' into espnow-packet-transport

bdraco committed:
Merge branch 'dev' into espnow-packet-transport

swoboda1337 committed:
Remove static

swoboda1337 committed:
Always register broadcast

swoboda1337 committed:
Fix comment
